### PR TITLE
add migration for django 4, fix serializer validation

### PR DESCRIPTION
**Pull request by gsnider2195**

### CODE REVIEW

File: nautobot_firewall_models/tests/test_api.py [hunk 1]
```diff
@@ -27,6 +31,67 @@ def setUpTestData(cls):
...
+        # Create an IPRange object with a vrf
+        models.IPRange.objects.create(start_address="1.0.0.1", end_address="1.0.0.8", vrf=vrfs[0])
+
+        initial_count = self._get_queryset().count()
+
+        # Create an IPRange object with the same start and end address but a different vrf
+        with disable_warnings("django.request"):
+            data = {"start_address": "1.0.0.1", "end_address": "1.0.0.8", "vrf": vrfs[1].pk}
+            response = self.client.post(url, data, format="json", **self.header)
+            self.assertHttpStatus(response, status.HTTP_201_CREATED)
```
Comment (timizuoebideri1) [inline nautobot_firewall_models/tests/test_api.py:58]:
```suggestion
            self.assertHttpStatus(response, rf_status.HTTP_201_CREATED)
```

File: nautobot_firewall_models/tests/test_api.py [hunk 2]
```diff
@@ -27,6 +31,67 @@ def setUpTestData(cls):
...
+            self.assertEqual(self._get_queryset().count(), initial_count + 1)
+
+        # Creating an IPRange object with the same start and end address and the same vrf fails
+        with disable_warnings("django.request"):
+            data = {"start_address": "1.0.0.1", "end_address": "1.0.0.8", "vrf": vrfs[0].pk}
+            response = self.client.post(url, data, format="json", **self.header)
+            self.assertHttpStatus(response, status.HTTP_400_BAD_REQUEST)
```
Comment (timizuoebideri1) [inline nautobot_firewall_models/tests/test_api.py:65]:
```suggestion
            self.assertHttpStatus(response, rf_status.HTTP_400_BAD_REQUEST)
```

File: nautobot_firewall_models/tests/test_api.py [hunk 3]
```diff
@@ -27,6 +31,67 @@ def setUpTestData(cls):
...
+            self.assertEqual(self._get_queryset().count(), initial_count + 1)
+
+        # Create another IPRange object with no vrf
+        models.IPRange.objects.create(start_address="2.0.0.1", end_address="2.0.0.8")
+
+        # Creating an IPRange object with the same start and end address with no vrf fails
+        with disable_warnings("django.request"):
+            data = {"start_address": "2.0.0.1", "end_address": "2.0.0.8"}
+            response = self.client.post(url, data, format="json", **self.header)
+            self.assertHttpStatus(response, status.HTTP_400_BAD_REQUEST)
```
Comment (timizuoebideri1) [inline nautobot_firewall_models/tests/test_api.py:80]:
```suggestion
            self.assertHttpStatus(response, rf_status.HTTP_400_BAD_REQUEST)
```

File: nautobot_firewall_models/tests/test_api.py [hunk 4]
```diff
@@ -27,6 +31,67 @@ def setUpTestData(cls):
...
+                "The fields start_address, end_address must make a unique set.",
+                response.data["non_field_errors"][0],
+            )
+            self.assertEqual(self._get_queryset().count(), initial_count + 2)
+
+        # Create an IPRange object with the same start and end address with a vrf
+        with disable_warnings("django.request"):
+            data = {"start_address": "2.0.0.1", "end_address": "2.0.0.8", "vrf": vrfs[0].pk}
+            response = self.client.post(url, data, format="json", **self.header)
+            self.assertHttpStatus(response, status.HTTP_201_CREATED)
```
Comment (timizuoebideri1) [inline nautobot_firewall_models/tests/test_api.py:92]:
```suggestion
            self.assertHttpStatus(response, rf_status.HTTP_201_CREATED)
```

File: nautobot_firewall_models/tests/test_api.py [hunk 5]
```diff
@@ -1,11 +1,15 @@
...
 # pylint: disable=duplicate-code
-from nautobot.apps.testing import APIViewTestCases
+from django.contrib.contenttypes.models import ContentType
+from nautobot.apps.testing import APIViewTestCases, disable_warnings
 from nautobot.dcim.models import Location, Platform, DeviceType, Device
 from nautobot.extras.models import Status, Role
-from nautobot.ipam.models import Prefix
+from nautobot.ipam.models import Prefix, VRF
+from nautobot.users.models import ObjectPermission
+from rest_framework import status
```
[the remark this answers is not in the document]
Comment (timizuoebideri1) [reply nautobot_firewall_models/tests/test_api.py:12]: PyLint: `nautobot_firewall_models/tests/test_api.py:432:8: W0621: Redefining name 'status' from outer scope (line 12) (redefined-outer-name)`

File: nautobot_firewall_models/migrations/0021_alter_addressobject_status_and_more.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+from django.db import migrations
+import django.db.models.deletion
+import nautobot.extras.models.statuses
+import nautobot_firewall_models.utils
+
+
+class Migration(migrations.Migration):
+    dependencies = [
+        ("extras", "0113_saved_views"),
```
Comment (glennmatthews) [inline nautobot_firewall_models/migrations/0021_alter_addressobject_status_and_more.py:11]: Is this migration backwards compatible with Django 3 or does this make Nautobot 2.3 the new required minimum version?

Comment (gsnider2195) [reply nautobot_firewall_models/migrations/0021_alter_addressobject_status_and_more.py]: It should be good now that I removed this dependency on extras 0113. It looks like the behavior change is only in the generation of migration files and in my testing these migrations are backward compatible.

**Review by timizuoebideri1 — APPROVED** [review]

**Review by glennmatthews — APPROVED** [review]

**Pull request merged**